File: New folder/cdp-sdk-8fcfac348487f586b224b77074f52594ad44198b/examples/typescript/evm/spend-permissions/account.useSpendPermission.ts
// Usage: pnpm tsx evm/spend-permissions/account.useSpendPermission.ts
import { CdpClient, parseUnits } from "@coinbase/cdp-sdk";

import "dotenv/config";

const cdp = new CdpClient();

const account = await cdp.evm.getOrCreateSmartAccount({
  name: "Demo-SpendPermissions-Account",
  owner: await cdp.evm.getOrCreateAccount({
    name: "Demo-SpendPermissions-Account-Owner",
  }),
  enableSpendPermissions: true,
});

const spender = await cdp.evm.getOrCreateAccount({
  name: "Demo-SpendPermissions-EOA-Spender",
});

const baseSpender = await spender.useNetwork("base-sepolia");

console.log("Account:", account.address);
console.log("Spender account:", spender.address);

if (process.argv.includes("--faucet")) {
  // Fund the spender with ETH for gas
  const faucet = await baseSpender.requestFaucet({
    token: "eth",
  });
  console.log("Faucet transaction sent:", faucet.transactionHash);

  const txReceipt = await baseSpender.waitForTransactionReceipt(faucet);
  console.log("Faucet transaction completed:", txReceipt.transactionHash);
}

console.log("Creating spend permission...");
const createSpendResult = await cdp.evm.createSpendPermission({
  network: "base-sepolia",
  spendPermission: {
    account: account.address,
    spender: spender.address,
    token: "usdc",
    allowance: parseUnits("0.01", 6),
    periodInDays: 1,
  },
});

const createSpendReceipt = await account.waitForUserOperation(
  createSpendResult
);
if (createSpendReceipt.status === "failed") {
  console.log(
    "Failed to create spend permission:",
    createSpendReceipt.userOpHash
  );
  process.exit(1);
}

console.log("Spend permission created");

// Sleep 2 seconds
await new Promise((resolve) => setTimeout(resolve, 2000));

// List spend permissions to get the actual resolved permission
const allPermissions = await cdp.evm.listSpendPermissions({
  address: account.address,
});
const permissions = allPermissions.spendPermissions.filter(
  (p) => p.permission.spender.toLowerCase() === spender.address.toLowerCase()
);

console.log("Executing spend...");

// Use the spend permission
const spend = await baseSpender.useSpendPermission({
  spendPermission: permissions.at(-1)!.permission, // Use the latest permission
  value: parseUnits("0.005", 6), // Spend 0.005 USDC (half the allowance)
});

console.log("Spend sent, waiting for receipt...");

const spendReceipt = await baseSpender.waitForTransactionReceipt(spend);

console.log("Spend completed!");
console.log(
  "Transaction:",
  `https://sepolia.basescan.org/tx/${spendReceipt.transactionHash}`
);

File: New folder/cdp-sdk-8fcfac348487f586b224b77074f52594ad44198b/examples/python/evm/spend-permissions/account.use_spend_permission.py
# Usage: uv run python evm/spend-permissions/account.use_spend_permission.py

import asyncio

from web3 import Web3

from cdp import CdpClient
from cdp.spend_permissions import SpendPermissionInput
from cdp.utils import parse_units

from dotenv import load_dotenv

load_dotenv()

web3 = Web3(Web3.HTTPProvider("https://sepolia.base.org"))


async def main():
    """Main function to demonstrate using a spend permission with a smart account."""
    async with CdpClient() as cdp:
        # Create accounts for the example
        master_owner = await cdp.evm.get_or_create_account(
            name="Demo-SpendPermissions-Master-Owner"
        )
        master = await cdp.evm.get_or_create_smart_account(
            name="Demo-SpendPermissions-Master",
            owner=master_owner,
            enable_spend_permissions=True,
        )

        spender = await cdp.evm.get_or_create_account(
            name="Demo-SpendPermissions-EOA-Spender"
        )

        faucet_tx_hash = await spender.request_faucet(
            network="base-sepolia", token="eth"
        )
        print(f"Faucet transaction sent: {faucet_tx_hash}")
        tx_receipt = web3.eth.wait_for_transaction_receipt(faucet_tx_hash)
        print(f"Faucet transaction completed: {tx_receipt.transactionHash}")

        print(f"Master account: {master.address}")
        print(f"Spender account: {spender.address}")

        spend_permission = SpendPermissionInput(
            account=master.address,
            spender=spender.address,
            token="usdc",
            allowance=parse_units("0.01", 6),
            period_in_days=1,
        )

        # Create the spend permission onchain
        user_operation = await cdp.evm.create_spend_permission(
            spend_permission=spend_permission,
            network="base-sepolia",
        )
        print(
            f"Created spend permission with user operation hash: {user_operation.user_op_hash}"
        )

        # Wait for the user operation to complete
        result = await cdp.evm.wait_for_user_operation(
            smart_account_address=master.address,
            user_op_hash=user_operation.user_op_hash,
        )
        print(f"User operation completed with status: {result.status}")

        # Sleep 2 seconds
        await asyncio.sleep(2)

        all_permissions = await cdp.evm.list_spend_permissions(master.address)
        permissions = [
            permission
            for permission in all_permissions.spend_permissions
            if permission.permission.spender == spender.address.lower()
        ]

        print("Executing spend...")

        # Use the spend permission
        spend_tx_hash = await spender.use_spend_permission(
            spend_permission=permissions[-1].permission,
            value=parse_units("0.005", 6),  # Spend 0.005 USDC (half the allowance)
            network="base-sepolia",
        )

        print(f"Spend sent, waiting for receipt... {spend_tx_hash}")

        tx_receipt = web3.eth.wait_for_transaction_receipt(spend_tx_hash)

        print("Spend completed!")
        print(
            f"Transaction: https://sepolia.basescan.org/tx/{tx_receipt.transactionHash.hex()}"
        )


if __name__ == "__main__":
    asyncio.run(main())
